### PR TITLE
hostpath topology + CSI ephemeral inline volume scheduling

### PR DESCRIPTION
**Pull request by pohly**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

It should have been possible to run a CSI driver only on some nodes in a cluster (or just one, as for the CSI hostpath example driver) and then use that driver for an CSI ephemeral inline volume without manually scheduling the pod. This hasn't worked so far because the pod scheduler was unaware of this limitation.

**Special notes for your reviewer**:

This builds on top of https://github.com/kubernetes/kubernetes/pull/88050 which failed because of that problem.

This is best reviewed commit-by-commit, because these are multiple different, logically independent steps.

**Does this PR introduce a user-facing change?**:
```release-note
Pods with CSI ephemeral inline volumes will only be scheduled onto nodes where the CSI driver for each volume is running. Beware that uninstalling a driver on a node may not be noticed immediately, so there's still a risk that pods incorrectly land on that node for a short while.
```

### CODE REVIEW

**Comment by pohly** [comment]
/assign @msau42 

I still need to add more tests for the new failure scenarios, but before I do that let me ask for some feedback whether these changes are okay in principle.

File: pkg/controller/volume/scheduling/scheduler_binder.go [hunk 1]
```diff
@@ -238,7 +245,7 @@ func (b *volumeBinder) FindPodVolumes(pod *v1.Pod, node *v1.Node) (unboundVolume
 			if selectedNode, ok := claim.Annotations[pvutil.AnnSelectedNode]; ok {
 				if selectedNode != node.Name {
 					// Fast path, skip unmatched node
-					return false, boundVolumesSatisfied, nil
+					return []string{ErrReasonBindConflict}, nil
```
Comment (pohly) [inline pkg/controller/volume/scheduling/scheduler_binder.go]: I need to point out one minor simplification: previously, both booleans might have been `false` here. Now only `ErrReasonBindConflict` is returned as reason for the scheduling failure.

None of the unit tests covered that particular combination. I think its enough to return one reason here.

File: pkg/scheduler/core/generic_scheduler_test.go [hunk 2]
```diff
@@ -37,6 +37,7 @@ import (
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/informers"
 	clientsetfake "k8s.io/client-go/kubernetes/fake"
+	"k8s.io/kubernetes/pkg/controller/volume/scheduling"
```
Comment (ahg-g) [inline pkg/scheduler/core/generic_scheduler_test.go]: please name the package ```volumescheduling```

**Review by ahg-g — COMMENTED** [review]
scheduler related changes looks ok, just a minor nit.

File: pkg/controller/volume/scheduling/scheduler_binder.go [hunk 3]
```diff
@@ -455,6 +482,35 @@ var (
 	versioner = etcd3.APIObjectVersioner{}
 )
 
+// checkInlineVolumes ensures that the CSI driver required by each volume
+// is really running on the node.
+func (b *volumeBinder) checkInlineVolumes(nodeName string, inlineVolumes []*v1.CSIVolumeSource) (bool, error) {
```
Comment (msau42) [inline pkg/controller/volume/scheduling/scheduler_binder.go]: This is going to break integration with cluster autoscaler. Cluster autoscaler works by simulating Node objects, and there is currently no way to tell it what nodes a CSI driver is expected to be installed on. The current expectation is that a CSI driver will be installed on all nodes. I think the current way we deploy csi hostpath only on a single node is more of an anti-pattern.

Comment (pohly) [reply pkg/controller/volume/scheduling/scheduler_binder.go]: > This is going to break integration with cluster autoscaler. Cluster autoscaler works by simulating Node objects, and there is currently no way to tell it what nodes a CSI driver is expected to be installed on.

So the autoscaler simulates nodes, a pod gets scheduled onto it, and then the autoscaler actually creates the node? Yes, that won't work for pods using ephemeral inline volumes with this check active.

> The current expectation is that a CSI driver will be installed on all nodes.

Sounds like something that we should document.

**Comment by pohly** [comment]
So 3de83473759a7ca39920cd178b148a5767458c67 is not going to work.

Is 0bae836c0c1a81f8ef9a2d6df5f5579786fff1d5 a useful refactoring that I should submit separately (or in this PR, after removing the rest)?

**Comment by msau42** [comment]
> Is [0bae836](https://github.com/kubernetes/kubernetes/commit/0bae836c0c1a81f8ef9a2d6df5f5579786fff1d5) a useful refactoring that I should submit separately (or in this PR, after removing the rest)?

Seems reasonable to me

**Comment by pohly** [comment]
> > Is 0bae836 a useful refactoring that I should submit separately (or in this PR, after removing the rest)?
>
> Seems reasonable to me

See https://github.com/kubernetes/kubernetes/pull/88230.

/close

**Pull request closed without merging**